### PR TITLE
Add player key index in database to speedup player lookups

### LINKED ISSUE #204: [BUG] Very slow lookup
**Describe the bug**
After a PR #194 some prism lookups began to run very slowly.
For example this command `/pr l r:100 t:1d -no-group` executed ~30s before PR and ~600s after on our prism database with 18kk+ rows.

This is because database engine using wrong key (player) when select.

For `/pr l r:100 t:1d -no-group`:

```EXPLAIN SELECT id, epoch, action_id, player, world_id, x, y, z, block_id, block_subid, old_block_id, old_block_subid, data, HEX(player_uuid) AS uuid FROM prism_data INNER JOIN prism_players p ON p.player_id = prism_data.player_id LEFT JOIN prism_data_extra ex ON ex.data_id = prism_data.id WHERE world_id = ( SELECT w.world_id FROM prism_worlds w WHERE w.world = 'world') AND (prism_data.x BETWEEN 1100 AND 1300) AND (prism_data.y BETWEEN 102 AND 302) AND (prism_data.z BETWEEN 2000 AND 2200) AND prism_data.epoch >= 1591875073 ORDER BY prism_data.epoch DESC, x ASC, z ASC, y ASC, id DESC LIMIT 1000```
slow

After adding `IGNORE INDEX (player)` to select query, the lookup is fast again.

```EXPLAIN SELECT id, epoch, action_id, player, world_id, x, y, z, block_id, block_subid, old_block_id, old_block_subid, data, HEX(player_uuid) AS uuid FROM prism_data IGNORE INDEX (player) INNER JOIN prism_players p ON p.player_id = prism_data.player_id LEFT JOIN prism_data_extra ex ON ex.data_id = prism_data.id WHERE world_id = ( SELECT w.world_id FROM prism_worlds w WHERE w.world = 'world') AND (prism_data.x BETWEEN 1100 AND 1300) AND (prism_data.y BETWEEN 102 AND 302) AND (prism_data.z BETWEEN 2000 AND 2200) AND prism_data.epoch >= 1591875073 ORDER BY prism_data.epoch DESC, x ASC, z ASC, y ASC, id DESC LIMIT 1000```
normal

So can anyone compare the query speed at their **large** database before pr and after?

**Server (please complete the following information):**
  - Prism Version: v2.1.7-SNAPSHOT-164
  - Server Type: Paper
  - Server Version: git-Paper-347
  - Database: MariaDB 10.2.29

### PR DESCRIPTION
**Pull request by lflare**
Not entirely sure if this conforms to standards and guidelines, but as I was doing some lookups the other day, I realised that all lookups involving players with global radius and insane times resulted in searches taking forever. This might help to alleviate that.

### CODE REVIEW

**Comment by Narimm** [comment]
This only adds the feature to new databases - we may need to consider a database structure update

**Comment by lflare** [comment]
I believe the functionality of database structure upgrades can be further explored down the line. Adding this PR will effectively only bring a performance boost without any unintended side-effects for those that have already created the database.

**Comment by addstar** [comment]
@lflare Great suggestion, thanks! I thought this was already implemented, I'm surprised it's not.

Any chance you could demonstrate the improvements with a comparative set of `EXPLAIN` queries and/or provide some timings? It certainly seems like a good thing to have, but I'm always hesitant about DB changes without data to justify the improvement.

**Comment by addstar** [comment]
Also, @Narimm / @lflare 
There's already a built-in schema updater (we're currently at schema version 7) and it's fairly simple to add new schema versions which will update existing schemas. Any schema change should have a corresponding schema update added, otherwise down the line users may face inconsistency issues with varying installs. I think it's just asking for trouble.

Add schema update here: https://github.com/AddstarMC/Prism-Bukkit/blob/master/src/main/java/me/botsko/prism/database/sql/SqlPrismDataSourceUpdater.java#L78

Bump version to 8 and call the update here: https://github.com/AddstarMC/Prism-Bukkit/blob/master/src/main/java/me/botsko/prism/Updater.java#L29

**Comment by Narimm** [comment]
I am aware of the schema updater...  it will need to be added before this is accepted

**Comment by lflare** [comment]
> @lflare Great suggestion, thanks! I thought this was already implemented, I'm surprised it's not.
> 
> Any chance you could demonstrate the improvements with a comparative set of `EXPLAIN` queries and/or provide some timings? It certainly seems like a good thing to have, but I'm always hesitant about DB changes without data to justify the improvement.

I am not currently able to provide such timings comparisons, but the comparisons can be simply replicated with `/pr l p:[PLAYER] t:1y r:global`, adding the index manually with ```ALTER TABLE data ADD KEY `player` (`player_id`);``` and re-running the query above with another player. This can then later be removed with ```ALTER TABLE DROP KEY `player`;```.

Either way, having an index on `player_id` will do far more good to lookup performance than harm. _The only issue I can see from this is the database server taking up just a tad more storage for the indexing._

**Comment by lflare** [comment]
Looks like I forgot entirely how to code. Once all the checks pass, I will squash all the commits together into one neat commit and force push it.

File: src/main/java/me/botsko/prism/database/sql/SqlPrismDataSourceUpdater.java [hunk 1]
```diff
@@ -91,4 +91,19 @@ public void v6_to_v7() {
             dataSource.handleDataSourceException(e);
         }
     }
+
+    @Override
+    public void v7_to_v8() {
+        String query;
+        try (
+            Connection conn = dataSource.getConnection();
+            PreparedStatement st = conn.prepareStatement(query);
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/database/sql/SqlPrismDataSourceUpdater.java:100]: Codacy Issue found: ['try resource' has incorrect indentation level 12, expected level should be 14.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620&bid=18379379)

File: src/main/java/me/botsko/prism/database/sql/SqlPrismDataSourceUpdater.java [hunk 2]
```diff
@@ -91,4 +91,19 @@ public void v6_to_v7() {
             dataSource.handleDataSourceException(e);
         }
     }
+
+    @Override
+    public void v7_to_v8() {
+        String query;
+        try (
+            Connection conn = dataSource.getConnection();
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/database/sql/SqlPrismDataSourceUpdater.java:99]: Codacy Issue found: ['try resource' has incorrect indentation level 12, expected level should be 14.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620)

File: src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java [hunk 3]
```diff
@@ -18,5 +18,6 @@
 
     void v6_to_v7();
 
+    void v7_to_v8();
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: Codacy Issue found: [The instance method name 'v7_to_v8' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620&bid=18379379)

File: src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java (hunk 3 above)
Comment (lflare) [reply src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: Uh. I'm fairly sure none of the other instance methods match that regular expression...

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620&bid=18379379)

File: src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java (hunk 3 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: Codacy Issue found: [The instance method name 'v7_to_v8' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620&bid=18379379)

File: src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java (hunk 3 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: Codacy Issue found: [The instance method name 'v7_to_v8' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620&bid=18379379)

File: src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java (hunk 3 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: Codacy Issue found: [The instance method name 'v7_to_v8' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620)

Comment (lflare) [reply src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: I believe the current naming is in line with the other function namings.

**Review by alchemistmatt — APPROVED** [review]
I approve of the addition of this index; it will definitely help query performance.  Good to see you will also auto-create it.  My only suggestion would be to name the index `ix_player_id` instead of just `player`, but the name doesn't matter much, and the other index is already just `location`

**Comment by Narimm** [comment]
To support the addition of database changes in SNAPSHOT releases . I would prefer that we dont version the db with every commit .  I am not quite sure yet how to handle snapshot db updates - because the problem is anyone who installs a snapshot build wont get any updates in the final release.    

My tendency is to enforce running the latest update in any snapshot build - so we need any table alter statements to handles being run multiple times on a potentitall already updated DB--- OR check the db first and skip.

**Comment by addstar** [comment]
> To support the addition of database changes in SNAPSHOT releases . I would prefer that we dont version the db with every commit . I am not quite sure yet how to handle snapshot db updates - because the problem is anyone who installs a snapshot build wont get any updates in the final release.

This should work the same way you'd do it in any production deployment with schema changes. You don't version the DB with each commit, you version the change to the actual schema itself. The code commit is irrelevant, what's important is the merge to master. If you merge a schema change to master, you should _always_ consider it final. Any further adjustment to the schema (even if you change your mind on how the previous schema change was done) should create a new version bump.

So an individual PR (not commit) should contain a single schema version bump and once it's merged, that version and schema change is committed and can never be changed.

> My tendency is to enforce running the latest update in any snapshot build - so we need any table alter statements to handles being run multiple times on a potentitall already updated DB--- OR check the db first and skip.

I don't really understand what you mean here, but IMO, we should just _always_ be enforcing all merged schema changes in all environments. We should not require anyone who uses a snapshot to do anything special, there's no need for it. When the schema update is merged and they run the snapshot with it, it will update their schema and the schema version in the DB so it will never attempt the update again.

**Comment by lflare** [comment]
> I approve of the addition of this index; it will definitely help query performance. Good to see you will also auto-create it. My only suggestion would be to name the index `ix_player_id` instead of just `player`, but the name doesn't matter much, and the other index is already just `location`

That's my intention, to follow the convention of whatever has been done prior.

> To support the addition of database changes in SNAPSHOT releases . I would prefer that we dont version the db with every commit . I am not quite sure yet how to handle snapshot db updates - because the problem is anyone who installs a snapshot build wont get any updates in the final release.
> 
> My tendency is to enforce running the latest update in any snapshot build - so we need any table alter statements to handles being run multiple times on a potentitall already updated DB--- OR check the db first and skip.

I reckon it will be quite simple to handle these kind of things, by having a value in the database that specifies the current database schema version, and from that, retroactively apply all the updates that need to be applied to match the database schema of the current.

With that in mind, I believe it is best practice to version the schema as long as it makes it into a public build, like a snapshot. If changes are made to the schema, that should be added to the history, irregardless of however minor it is. It might not be a bad idea to take a look at how Laravel handles database schema versioning too.

With all of that said and done, I do not believe that there is anything lacking in this PR?

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620&bid=18379379)

File: src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java (hunk 3 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: Codacy Issue found: [The instance method name 'v7_to_v8' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5609620)

Comment (lflare) [reply src/main/java/me/botsko/prism/database/PrismDataSourceUpdater.java:21]: Please confirm if you want this resolved.

**Pull request merged**